[Feature] Create Catchment

luongt13 opened the pull request:
Closes https://github.com/crosscutio/dhis2-crosscut-app/issues/5

This PR connects with the `catchment-jobs` endpoint to create a catchment. If the `csv` has errors, the user can remove the lines with errors. This PR also includes code to add info and the ability to delete from crosscut. The issues for info and delete will be addressed in different PRs. 

To address the 500 error in another PR: https://github.com/crosscutio/dhis2-crosscut-app/issues/20

briannapoulos commented on src/api/crosscutRequests.js:
```diff
@@ -16,11 +16,8 @@ export const fetchCatchmentJobs = async () => {
-        const catchmentsPublished = await fetchCurrentAttributes()
 
-        // TODO: check for published catchments to update the status
+         // TODO: check for published catchments to update the status
```
it looks like an extra space was added here

briannapoulos commented on src/api/crosscutRequests.js:
```diff
@@ -30,14 +27,16 @@ export const fetchCatchmentJobs = async () => {
         const siteBasedJobs = resp.jobs.filter((job) => job.algorithm === "site-based")
-        
+
+        // TODO: sort by time it was created
```
I think it would be better to sort by the catchment ID, since the catchment ID is how we sort it in the app (and sorting it by catchment ID sorts it by the date created)

briannapoulos commented on src/api/crosscutRequests.js:
```diff
@@ -95,4 +96,30 @@ export const deleteCatchmentJob = async (id) => {
+        }).json()
+
+        resp.boundaryList.forEach((bound) => {
+            if (bound.areaName === "" && bound.entireCountry === true) {
+                bound.areaName = i18n.t("entire country")
+            }
```
For the moment, do we only want to include entire countries in the drop-down list? Will we be able to support partial countries before https://github.com/crosscutio/dhis2-crosscut-app/issues/9 is complete?

luongt13 replied:
A user should be able to create catchments as they will receive errors on the sites that fall outside of our boundary and create the catchment with the rest of the sites.

briannapoulos commented on src/components/Create/Create.jsx:
```diff
@@ -134,12 +231,54 @@ function Create(props) {
+            {hasErrors && <Modal>
+                <ModalTitle>{errorData.errors.length} {i18n.t("errors were found")}</ModalTitle>
+                <ModalContent>
+                    {i18n.t("Click proceed if you want to continue and the errors will automatically be removed or click cancel to go back.")}
```
I think we should clarify that the sites with errors will be removed (not just the errors themselves)

briannapoulos reviewed (changes requested):
I tried to create a catchment for SWZ (not really expecting it to pass, since I think only SLE works at the moment) and it caused an error in the backend but the error doesn't look like it propagated to the app (the Create button just shows loading indefinitely). I think we should let the user know if there is an error returned when attempting to create the catchment.
<img width="583" alt="Screen Shot 2022-04-21 at 3 53 33 PM" src="https://user-images.githubusercontent.com/13878539/164541969-1f1e5459-2a1f-4451-9956-78474e7ea99e.png">

luongt13 commented:
> I tried to create a catchment for SWZ (not really expecting it to pass, since I think only SLE works at the moment) and it caused an error in the backend but the error doesn't look like it propagated to the app (the Create button just shows loading indefinitely). I think we should let the user know if there is an error returned when attempting to create the catchment. 

I've added a check to see if there are no valid sites (no error) then the user can't proceed. I've added an alert bar for other errors (i.e., internal errors).

briannapoulos commented on i18n/en.pot:
```diff
@@ -17,12 +17,47 @@ msgstr "Publish"
+
+msgid ""
+"Click proceed if you want to continue. The sites with errors will be "
+"removed if you proceed or click cancel to go back."
```
I think it would be good to have a comma after `proceed` to clarify that the sites with errors won't be removed if you click cancel to go back.

briannapoulos commented on src/components/Create/Create.jsx:
```diff
@@ -81,42 +153,93 @@ function Create(props) {
+        
+        // if there are errors then set the error
+        if (resp?.error) {
+            const errors = resp.error.data.filter((data) => data["cc:ErrorMessage"] !== "") 
+            const clean = resp.error.data.filter((data) => data["cc:ErrorMessage"] === "")
```
If it's easy to modify, I think it would be better to show the user a column with the title "ERROR MESSAGE" instead of "cc:ErrorMessage" -- having the "cc:" makes it look out of place with the other column headings.

briannapoulos commented on src/components/Create/Create.jsx:
```diff
@@ -81,42 +153,93 @@ function Create(props) {
+            // toggle to fetch for jobs
+            toggle()
+            setIsLoading(false)
+            setAlert({ text: i18n.t("Your catchment areas are being created. It should be ready in a few minutes")})
```
I think `It` should be changed to `They` since `areas` is plural in the prior sentence.

luongt13 replied:
This was taken from our application so we can use the translation from there. It's a little confusing because a catchment job has catchment areas and it is referring to the catchment job.

briannapoulos reviewed (changes requested):
Thank you for addressing the comments! It was cool to see the catchment creation succeed when triggered by DHIS2.

I had a couple other minor suggestions. Thanks!
